### PR TITLE
Use macros to manage any domain id and effective domain id related operations

### PR DESCRIPTION
**Pull request by thyagarajanqc**
Use macros to manage any domain id and effective domain id related operations and convert listener info array tohas-table

### CODE REVIEW

File: inc/fastrpc_hash_table.h [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	typedef struct { \
+		type *tbl; \
+		pthread_mutex_t mut; \
+	} name##_table; \
+	static name##_table info;
+
+/* Initialize hash-table and associated members */
+#define HASH_TABLE_INIT(type) \
+	do {\
+		pthread_mutex_init(&info.mut, 0); \
```
Comment (quic-mtharu) [inline inc/fastrpc_hash_table.h:25]: It’s good practice to check the return value of pthread_mutex_init to ensure the mutex is initialized successfully.


    do { \
        if (pthread_mutex_init(&info.mut, NULL) != 0) { \
            perror("Mutex initialization failed"); \
            exit(EXIT_FAILURE); \
        } \
    } while(0)

File: inc/fastrpc_hash_table.h [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+	} while(0)
+
+/* Delete & all entries in hash-table */
+#define HASH_TABLE_CLEANUP(type) \
+	do { \
+		type *me = NULL, *tmp = NULL; \
+		\
+		pthread_mutex_lock(&info.mut); \
```
Comment (quic-mtharu) [inline inc/fastrpc_hash_table.h:33]: Similarly, check the return values of pthread_mutex_lock, pthread_mutex_unlock and pthread_mutex_destroy to handle any potential errors.

        if (pthread_mutex_lock(&info.mut) != 0) { \
            perror("Mutex lock failed"); \
            exit(EXIT_FAILURE); \
        } \
        
        if (pthread_mutex_unlock(&info.mut) != 0) { \
            perror("Mutex unlock failed"); \
            exit(EXIT_FAILURE); \
        } \

        if (pthread_mutex_destroy(&info.mut) != 0) { \
            perror("Mutex destruction failed"); \
            exit(EXIT_FAILURE); \
        } \

File: inc/fastrpc_hash_table.h [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+/* Declare hash-table struct and variable with given name */
+#define DECLARE_HASH_TABLE(name, type) \
+	typedef struct { \
+		type *tbl; \
+		pthread_mutex_t mut; \
+	} name##_table; \
+	static name##_table info;
+
+/* Initialize hash-table and associated members */
+#define HASH_TABLE_INIT(type) \
```
Comment (quic-mtharu) [inline inc/fastrpc_hash_table.h:23]: the "type" parameter is not used. You might want to remove it if it’s unnecessary.

File: inc/fastrpc_hash_table.h (hunk 1 above)
Comment (thyagarajanqc) [reply inc/fastrpc_hash_table.h:25]: Will fix after all subsequent changes are merged. Otherwise, there will be major merge conflicts.

File: inc/fastrpc_hash_table.h (hunk 3 above)
Comment (thyagarajanqc) [reply inc/fastrpc_hash_table.h:23]: Will fix after all subsequent changes are merged. Otherwise, there will be major merge conflicts.

File: inc/fastrpc_hash_table.h (hunk 2 above)
Comment (thyagarajanqc) [reply inc/fastrpc_hash_table.h:33]: Will fix after all subsequent changes are merged. Otherwise, there will be major merge conflicts.

File: src/listener_android.c [hunk 4]
```diff
@@ -465,9 +469,11 @@ int close_reverse_handle(remote_handle64 h, char *dlerr, int dlerrorLen,
 }
 
 int listener_android_geteventfd(int domain, int *fd) {
-  struct listener *me = &linfo[domain];
+  struct listener* me = NULL;
```
Comment (quic-mtharu) [inline src/listener_android.c]: Could we update the pointer declaration style for consistency? Instead of struct listener* me, let’s use struct listener *me to match our coding standards.

**Review by quic-mtharu — APPROVED** [review]

File: inc/fastrpc_hash_table.h [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+#define ADD_DOMAIN_HASH() \
+	int domain; \
+	UT_hash_handle hh;
+
+/* Declare hash-table struct and variable with given name */
+#define DECLARE_HASH_TABLE(name, type) \
+	typedef struct { \
+		type *tbl; \
+		pthread_mutex_t mut; \
+	} name##_table; \
```
Comment (quic-mtharu) [inline inc/fastrpc_hash_table.h:19]: It’s a good practice to use all uppercase letters for macro names to distinguish them from regular variables and functions.

Change **info** to **NAME##_info** to ensure the variable name is unique and associate with the macro’s NAME

    #define DECLARE_HASH_TABLE(NAME, TYPE) \
    typedef struct { \
        TYPE *tbl; \
        pthread_mutex_t mut; \
    } NAME##_table; \
    static NAME##_table NAME##_info;

Comment (thyagarajanqc) [reply inc/fastrpc_hash_table.h:19]: But this is not a macro. It is a struct and a global variable.

Comment (quic-mtharu) [reply inc/fastrpc_hash_table.h:19]: Your macro definition for declaring a hash-table struct and variable looks mostly correct, but there are a few points to consider for clarity and potential improvements:

**Macro Naming Convention:** if we use uppercase letters for macro names to distinguish them from regular variables and functions, can make macros more readable and maintainable, even when using token concatenation.
Following a consistent naming convention across a codebase makes it easier for teams to collaborate. It sets clear expectations and reduces the cognitive load when switching between different parts of the code.
**Uppercase Macro Name**: Change **name** and **type** to NAME and TYPE to follow the convention.

**Variable Naming:** Changed info to NAME##_info to ensure the variable name is unique and associated with the macro’s NAME. This approach ensures that the macro is flexible and can be used with different hash table names without hardcoding the variable name inside the macro.

File: src/apps_mem_imp.c [hunk 6]
```diff
@@ -30,9 +31,14 @@
 #define FASTRPC_ALLOC_HLOS_FD                                                  \
   0x10000 /* Flag to allocate HLOS FD to be shared with DSP */
 
-static QList memlst[NUM_DOMAINS_EXTEND];
-static pthread_mutex_t memmt[NUM_DOMAINS_EXTEND];
-int mem_init_flag[NUM_DOMAINS_EXTEND];
+typedef struct {
+	QList memlst;
```
Comment (quic-mtharu) [inline src/apps_mem_imp.c]: memlst -> mem_list
memmt -> mem_mut

File: src/fastrpc_notif.c [hunk 7]
```diff
@@ -28,11 +28,13 @@
 #include "fastrpc_notif.h"
 #include "platform_libs.h"
 #include "verify.h"
+#include "fastrpc_hash_table.h"
 
 struct notif_config {
```
Comment (quic-mtharu) [inline src/fastrpc_notif.c]: Can we replace the named struct _struct notif_config {}_ with a typedef struct _typedef struct {} notif_config;_? This change will align with the pattern we’re using for the apps_mem_info struct, making our code more consistent and easier to understand.

File: src/listener_android.c [hunk 8]
```diff
@@ -32,18 +32,19 @@
 #include "rpcmem.h"
 #include "shared.h"
 #include "verify.h"
+#include "fastrpc_hash_table.h"
 
 struct listener {
```
Comment (quic-mtharu) [inline src/listener_android.c]: Can we replace the **Named Struct** struct listener {} with a **Typedef Struct** typedef struct {} listener;? This change will align with the pattern we’re using for the apps_mem_info struct, making our code more consistent and easier to understand.

File: inc/fastrpc_internal.h [hunk 9]
```diff
@@ -77,12 +77,24 @@ static __inline uint32 Q6_R_cl0_R(uint32 num) {
 /* From actual domain ID (0-3) and session ID, get effective domain ID */
 #define GET_EFFECTIVE_DOMAIN_ID(domain, session) (domain + (NUM_DOMAINS * session))
 
+/* From effective domain ID, get actual domain ID */
+#define GET_DOMAIN_FROM_EFFEC_DOMAIN_ID(effec_dom_id) (effec_dom_id & DOMAIN_ID_MASK)
```
Comment (ekanshibu) [inline inc/fastrpc_internal.h:81]: this can be moved to a separate PR

File: src/listener_android.c [hunk 10]
```diff
@@ -469,7 +469,7 @@ int close_reverse_handle(remote_handle64 h, char *dlerr, int dlerrorLen,
 }
 
 int listener_android_geteventfd(int domain, int *fd) {
-  struct listener* me = NULL;
+  struct listener *me = NULL;
```
Comment (ekanshibu) [inline src/listener_android.c]: This also looks independent of multi-core, please move to a separate PR or squash this with listener hash table change

File: src/apps_mem_imp.c (hunk 6 above)
Comment (ekanshibu) [inline src/apps_mem_imp.c]: can you take all hast-table changes into a separate PR and we can plan to merge it before multi-core changes

File: src/listener_android.c [hunk 11]
```diff
@@ -406,7 +405,11 @@ int listener_android_init(void) {
 }
 
 void listener_android_domain_deinit(int domain) {
-  struct listener *me = &linfo[domain];
+  struct listener* me = NULL;
```
Comment (ekanshibu) [inline src/listener_android.c]: struct listener *me = NULL;

File: src/fastrpc_context.c [hunk 12]
```diff
@@ -0,0 +1,334 @@
...
+		domain = ctx->effec_domain_ids[i];
+
+		// If no session was opened on domain, continue
+		if (!ctx->devs[i])
+			continue;
+
+		fastrpc_session_close(domain);
+	}
+	if (ctx->devs)
+		free(ctx->devs);
```
Comment (ekanshibu) [inline src/fastrpc_context.c]: freeing NULL pointer should be safe
"If ptr is NULL, no operation is performed."
https://man7.org/linux/man-pages/man3/malloc.3.html#DESCRIPTION

File: src/fastrpc_context.c [hunk 13]
```diff
@@ -0,0 +1,334 @@
...
+
+	// Basic sanity checks on client inputs
+	VERIFYC(create->effec_domain_ids && !create->flags, AEE_EBADPARM);
+	VERIFYC(num_domain_ids && num_domain_ids < NUM_DOMAINS_EXTEND,
+		AEE_EBADPARM);
+
+	VERIFYC(NULL != (ctx = fastrpc_context_init(num_domain_ids)),
+		AEE_EBADCONTEXT);
+	VERIFYC(AEE_SUCCESS == (nErr = fastrpc_context_init_domains(create,
+										ctx)), AEE_EBADPARM);
```
Comment (ekanshibu) [inline src/fastrpc_context.c]: indentation

File: src/dspqueue/dspqueue_cpu.c [hunk 14]
```diff
@@ -1964,12 +2342,54 @@ AEEResult dspqueue_read(dspqueue_t queue, uint32_t *flags, uint32_t max_buffers,
...
+	// Only one multi-domain stats request at a time.
+	pthread_mutex_lock(&q->mutex);
+	locked = true;
+
+	/*
+	 * Always return stats from first queue only as this api is currently
+	 * supported for single-domain queues only.
+	 */
+	VERIFY(AEE_SUCCESS == (nErr = dspqueue_get_stat(mdq->queues[0],
+									stat, value)));
```
Comment (ekanshibu) [inline src/dspqueue/dspqueue_cpu.c]: indentation

File: src/dspqueue/dspqueue_cpu.c [hunk 15]
```diff
@@ -1902,15 +2270,25 @@ AEEResult dspqueue_read(dspqueue_t queue, uint32_t *flags, uint32_t max_buffers,
...
 
+  if (q->mdq.is_mdq) {
+    // Recursively call 'dspqueue_read_noblock' on individual queues
+    return dspqueue_multidomain_read(q, flags, max_buffers,
+                  num_buffers, buffers, max_message_length,
+                  message_length, message, timeout_us, true, true);
+  }
+
+  pq = &q->header->resp_queue;
+  read_state = (struct dspqueue_packet_queue_state*) (((uintptr_t)q->header)
```
Comment (ekanshibu) [inline src/dspqueue/dspqueue_cpu.c]: space after dspqueue_packet_queue_state

File: src/dspqueue/dspqueue_cpu.c [hunk 16]
```diff
@@ -1902,15 +2270,25 @@ AEEResult dspqueue_read(dspqueue_t queue, uint32_t *flags, uint32_t max_buffers,
...
+                          + pq->read_state_offset);
+  wait_count = (_Atomic uint32_t*) &read_state->wait_count;
```
Comment (ekanshibu) [inline src/dspqueue/dspqueue_cpu.c]: space after  uint32_t

File: src/dspqueue/dspqueue_cpu.c [hunk 17]
```diff
@@ -1620,26 +1977,37 @@ AEEResult dspqueue_read_noblock(dspqueue_t queue, uint32_t *flags,
...
+                num_buffers, buffers, max_message_length,
+                message_length, message, 0, true, false);
+    }
+
+  dq = queues->domain_queues[q->domain];
+  pq = &q->header->resp_queue;
+  qp = (volatile const uint8_t*) (((uintptr_t)q->header) + pq->queue_offset);
+  read_state = (struct dspqueue_packet_queue_state*) (((uintptr_t)q->header)
+                        + pq->read_state_offset);
+  write_state = (struct dspqueue_packet_queue_state*) (((uintptr_t)q->header)
```
Comment (ekanshibu) [inline src/dspqueue/dspqueue_cpu.c]: space after dspqueue_packet_queue_state

File: src/dspqueue/dspqueue_cpu.c [hunk 18]
```diff
@@ -1620,26 +1977,37 @@ AEEResult dspqueue_read_noblock(dspqueue_t queue, uint32_t *flags,
...
+        // Recursively call 'dspqueue_read_noblock' on individual queues
+        return dspqueue_multidomain_read(q, flags, max_buffers,
+                num_buffers, buffers, max_message_length,
+                message_length, message, 0, true, false);
+    }
+
+  dq = queues->domain_queues[q->domain];
+  pq = &q->header->resp_queue;
+  qp = (volatile const uint8_t*) (((uintptr_t)q->header) + pq->queue_offset);
+  read_state = (struct dspqueue_packet_queue_state*) (((uintptr_t)q->header)
```
Comment (ekanshibu) [inline src/dspqueue/dspqueue_cpu.c]: space after dspqueue_packet_queue_state

File: src/dspqueue/dspqueue_cpu.c [hunk 19]
```diff
@@ -1620,26 +1977,37 @@ AEEResult dspqueue_read_noblock(dspqueue_t queue, uint32_t *flags,
...
+if (q->mdq.is_mdq) {
+        // Recursively call 'dspqueue_read_noblock' on individual queues
+        return dspqueue_multidomain_read(q, flags, max_buffers,
+                num_buffers, buffers, max_message_length,
+                message_length, message, 0, true, false);
+    }
+
+  dq = queues->domain_queues[q->domain];
+  pq = &q->header->resp_queue;
+  qp = (volatile const uint8_t*) (((uintptr_t)q->header) + pq->queue_offset);
```
Comment (ekanshibu) [inline src/dspqueue/dspqueue_cpu.c]: space after uint8_t

File: src/dspqueue/dspqueue_cpu.c [hunk 20]
```diff
@@ -1525,15 +1873,24 @@ AEEResult dspqueue_peek(dspqueue_t queue, uint32_t *flags,
...
   struct timespec ts;
 
+  if (q->mdq.is_mdq) {
+    // Recursively call 'dspqueue_read_noblock' on individual queues
+    return dspqueue_multidomain_read(q, flags, 0, num_buffers,
+            NULL, 0, message_length, NULL, timeout_us, false, true);
+  }
+
+  pq = &q->header->resp_queue;
+  read_state =(struct dspqueue_packet_queue_state*) (((uintptr_t)q->header)
```
Comment (ekanshibu) [inline src/dspqueue/dspqueue_cpu.c]: space after dspqueue_packet_queue_state

File: src/fastrpc_apps_user.c [hunk 21]
```diff
@@ -357,8 +362,10 @@ int fastrpc_session_open(int domain, int *dev) {
   return AEE_ECONNREFUSED;
 }
 
-int fastrpc_session_close(int domain, int dev) {
-  if (hlist[domain].dev == -1)
+int fastrpc_session_close(int domain) {
+  int dev = hlist[domain].dev;
+
+  if (dev == -1)
```
Comment (quic-mtharu) [inline src/fastrpc_apps_user.c]: here you are attempts to close **if dev == -1** ?, which is unnecessary and incorrect.

we should close only if the **dev >= 0** right ?, please check once.

File: src/listener_android.c (hunk 11 above)
Comment (quic-mtharu) [inline src/listener_android.c]: struct listener *me = NULL;

File: inc/fastrpc_internal.h (hunk 9 above)
Comment (thyagarajanqc) [reply inc/fastrpc_internal.h:81]: It is related to multi-core fastrpc enablement as the mechanism to extract domain id from effective domain id will change with dynamic core discovery.

File: src/listener_android.c (hunk 10 above)
Comment (thyagarajanqc) [reply src/listener_android.c]: Is it important to move small cleanups like this to separate PR ? It will be very hard to split the commits into separate pull requests. I had to spend a lot of time rebasing my changes on top of the opensource branch. 

If there is any easy way to do it, please let me know.

File: src/apps_mem_imp.c (hunk 6 above)
Comment (thyagarajanqc) [reply src/apps_mem_imp.c]: Same response as above.

File: src/listener_android.c (hunk 11 above)
Comment (thyagarajanqc) [reply src/listener_android.c]: Done

File: src/fastrpc_context.c (hunk 12 above)
Comment (thyagarajanqc) [reply src/fastrpc_context.c]: Done

File: src/fastrpc_context.c (hunk 13 above)
Comment (thyagarajanqc) [reply src/fastrpc_context.c]: Done

File: src/dspqueue/dspqueue_cpu.c (hunk 14 above)
Comment (thyagarajanqc) [reply src/dspqueue/dspqueue_cpu.c]: Done

File: src/dspqueue/dspqueue_cpu.c (hunk 15 above)
Comment (thyagarajanqc) [reply src/dspqueue/dspqueue_cpu.c]: Done

File: src/dspqueue/dspqueue_cpu.c (hunk 16 above)
Comment (thyagarajanqc) [reply src/dspqueue/dspqueue_cpu.c]: Done

File: src/dspqueue/dspqueue_cpu.c (hunk 17 above)
Comment (thyagarajanqc) [reply src/dspqueue/dspqueue_cpu.c]: Done

File: src/dspqueue/dspqueue_cpu.c (hunk 18 above)
Comment (thyagarajanqc) [reply src/dspqueue/dspqueue_cpu.c]: Done

File: src/dspqueue/dspqueue_cpu.c (hunk 19 above)
Comment (thyagarajanqc) [reply src/dspqueue/dspqueue_cpu.c]: Done

File: src/dspqueue/dspqueue_cpu.c (hunk 20 above)
Comment (thyagarajanqc) [reply src/dspqueue/dspqueue_cpu.c]: Done

File: src/apps_mem_imp.c (hunk 6 above)
Comment (thyagarajanqc) [reply src/apps_mem_imp.c]: Done

File: src/fastrpc_notif.c (hunk 7 above)
Comment (thyagarajanqc) [reply src/fastrpc_notif.c]: Done

File: src/listener_android.c (hunk 8 above)
Comment (thyagarajanqc) [reply src/listener_android.c]: Done

File: src/fastrpc_apps_user.c (hunk 21 above)
Comment (thyagarajanqc) [reply src/fastrpc_apps_user.c]: Done

File: src/fastrpc_context.c [hunk 22]
```diff
@@ -0,0 +1,331 @@
+/*
+  Copyright (c) 2024 Qualcomm Technologies, Inc.
+  All Rights Reserved.
+  Confidential and Proprietary - Qualcomm Technologies, Inc.
```
Comment (quic-mtharu) [inline src/fastrpc_context.c]: Could you please change the copyright to the format below?

// Copyright (c) 2024, Qualcomm Innovation Center, Inc. All rights reserved.
// SPDX-License-Identifier: BSD-3-Clause

File: inc/fastrpc_hash_table.h [hunk 23]
```diff
@@ -0,0 +1,70 @@
+/*
```
Comment (quic-mtharu) [inline inc/fastrpc_hash_table.h]: Could you please change the copyright to the opensource copyright format ?

File: inc/fastrpc_context.h [hunk 24]
```diff
@@ -0,0 +1,100 @@
+/*
```
Comment (quic-mtharu) [inline inc/fastrpc_context.h]: Could you please change the copyright to the opensource copyright format ?

**Review by quic-mtharu — APPROVED** [review]

**Review by quic-ekangupt — APPROVED** [review]
